consensus/istanbul/validator/random/random.go

```diff
@@ -0,0 +1,60 @@
+// Package random implements a language independent method of producing random validator orderings.
+// It is designed to use standard methods that could be implemented in any language or platform.
+package random
+
+import (
+	"encoding/binary"
+
+	"github.com/ethereum/go-ethereum/common"
+	"golang.org/x/crypto/sha3"
+)
+
+const (
+	// Arbitrary locally unique salts to prevent collisions.
+	uniformSalt     = 0xe0
+	permutationSalt = 0xe1
+)
+
+// Permutation produces an array with a random permutation of [0, 1, ... n-1]
+// Based on the Fisher-Yates method https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+func Permutation(randomness common.Hash, n int) []int {
+	if n <= 0 {
+		return nil
+	}
+
+	// Create the unshuffled array [0, 1, ... n-1]
+	array := make([]int, n)
+	for i := 0; i < n; i++ {
+		array[i] = i
+	}
+
+	// Shuffle the array using the Fisher-Yates method.
+	for i := 0; i < n-1; i++ {
+		randomness = sha3.Sum256(append(randomness[:], permutationSalt))
```

Comment:
This might be fine but it's going to be cleaner and more secure to use a stream cipher here like salsa20: https://godoc.org/golang.org/x/crypto/salsa20#XORKeyStream
You can just use the randomness passed into each function as a key along with a nonce that you increment as a counter, with the input being all 0s or some fixed seed phrase

Reply:
Switched to using SHAKE 256 in 8aca95d, much nicer!

Reply:
SHAKE256 is fine here but will be slower than a stream cipher if efficiency is a concern. Interface is nice though